Download beta compiler toolchain in bootstrap if it doesn't yet exist

onur-ozkan opened the pull request:
Blocker for #107812 and #99989

See: https://github.com/rust-lang/rust/pull/107812#discussion_r1143492202

r? @jyn514

jyn514 commented:
Feels a little not in the spirit of things for me to review my own commit 😆

r? bootstrap

Mark-Simulacrum commented on src/bootstrap/download.rs:
```diff
@@ -367,26 +367,69 @@ impl Config {
+
+    fn download_toolchain(
+        &self,
+        channel: &str,
```
On stable, this channel is going to look like "1.67.1" or similar. So it's closer to *version*, not channel.

Is there a reason not to pass the config::CompilerMetadata struct into download_toolchain?

ozkanonur replied:
> On stable, this channel is going to look like "1.67.1" or similar. So it's closer to version, not channel.

Yes, I see it's already passed as variable called `version` from other functions.

> Is there a reason not to pass the config::CompilerMetadata struct into download_toolchain?

`config::CompilerMetadata` isn't always used for passing `version`. In `download_ci_rustc`, `artifact_version_part` is used for calculating `version` which gets passed into `download_toolchain`.

Mark-Simulacrum replied:
I think it would be nice to refactor to always pass in CompilerMetadata, since it seems like we pass both of those fields here. That would also avoid having three &str arguments side-by-side which can be easy to flip accidentally (and I'm not sure we have good test coverage, so we might not catch that flip).

ozkanonur replied:
I fixed the previous review notes because they were pretty simple. Do you want me to add `FIXME` on top of the `version` arg, or do it in this PR? It requires some effort and I will not have enough space on this task for 3-4 days.

Mark-Simulacrum replied:
I would prefer to do it in this PR, but no strong feelings. r=me if you want to move ahead with a FIXME.

ozkanonur replied:
Let's not block this PR for that reactoring.

Just left a `FIXME(ozkanonur)` for it.

Mark-Simulacrum commented on src/bootstrap/config.rs:
```diff
@@ -223,36 +223,41 @@ pub struct Config {
 }
 
-#[derive(Default, Deserialize)]
-#[cfg_attr(test, derive(Clone))]
```
The unconditional Clone derive seems fine, but I can't see what causes it to be necessary. What is the cause for the change to be bundled here?

ozkanonur replied:
This change looks revertable since `Config` still keeps it as conditional.

Mark-Simulacrum commented:
@rustbot author

ozkanonur commented:
@bors r=Mark-Simulacrum rollup

bors commented:
:pushpin: Commit 6f8e47092cfbfbc944db848928bc92ca982041ea has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 6f8e47092cfbfbc944db848928bc92ca982041ea -->
<!-- homu: {"type":"Approved","sha":"6f8e47092cfbfbc944db848928bc92ca982041ea","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ehuss commented:
@bors r-

Failed in rollup: https://github.com/rust-lang/rust/pull/109872#issuecomment-1493535447

ozkanonur commented:
We are having problem at https://github.com/ozkanonur/rust/blob/6f8e47092cfbfbc944db848928bc92ca982041ea/src/bootstrap/download.rs#L421-L423 due to race condition on Windows platform.

I don't really want to add another dependency for Windows in bootstrap:( What if we don't remove the `bin_root` and leave it as it is, can't' overriding files and dirs work here?

for extra information, see:
- https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/When.20can.20.60remove_dir_all.60.20fail.20on.20Windows.3F/near/250010930
- https://github.com/rust-lang/rust/issues/29497

ChrisDenton commented:
I'm not sure that adding another dependency would help. This looks like a more fundamental issue. A process in the directory is still running (maybe bootstrap itself?) and this will prevent deletion.

ozkanonur commented:
Oh, I thought completely different stuff.

By roughly looking on the changes, the problem is seems like we are trying to remove `stage0` while using it.

ozkanonur commented on src/bootstrap/download.rs:
```diff
@@ -367,26 +367,74 @@ impl Config {
+        extra_components: &[&str],
+        download_component: fn(&Config, String, &str, &str),
+    ) {
+        if env::var_os("BOOTSTRAP_PYTHON").is_some() {
+            return
+        }
```
This PR forms a part of the `remove/replace python scripts from bootstrap`. Its objective is to download stage0 directly from the Rust source, rather than through the Python script. However, as we are still utilizing Python scripts currently, stage0 is downloaded twice. The second attempt involves executing the downloaded stage0 compiler, which is incompatible with Windows FS and causes an error, as can be seen in [this error](https://github.com/rust-lang-ci/rust/actions/runs/4591544753/jobs/8107789452#step:26:30881).


This check allows us to determine if bootstrap python involved or not(which also means stage0 is already downloaded). So we can prevent the removal of directories/files that are currently in use.

@jyn514 Please correct me if I am wrong since you are the author of this implementation.

jyn514 replied:
I don't think this check is correct. Imagine that we implement https://github.com/rust-lang/rust/issues/99989 in the python script. Then python will download only bootstrap-shim and not stage0, but `BOOTSTRAP_PYTHON` will still be set, so we'll get an error when we try to use stage0 to build std.

I would *expect* us to automatically avoid downloading the compiler if it already exists on disk, because of the `rustc_stamp` check. I think there is some different between the python stamp and the rust stamp on disk. Can you check and see if it's overwriting an existing stamp, maybe by adding `println!("{}", std::fs::read_to_string(rustc_stamp))` on line 420?

ozkanonur replied:
Sometimes we don't even create stamp file while bootstrap tests are performed.

I just discovered comparing stamp doesn't do anything wrong, checking the `rustc` executable what makes the tests fail. Is there any specific case that requires `!bin_root.join("bin").join("rustc").exists()` check? Tested on windows, when I remove this check from the if condition, all goes well.

ozkanonur replied:
LOL, it's because we are missing `.exe` extension on windows :cold_face:

jyn514 commented on src/bootstrap/download.rs:
```diff
@@ -367,26 +367,72 @@ impl Config {
         let rustc_stamp = bin_root.join(".rustc-stamp");
 
-        if !bin_root.join("bin").join("rustc").exists() || program_out_of_date(&rustc_stamp, commit)
+        let rustc_executable = if cfg!(target_os = "windows") { "rustc.exe" } else { "rustc" };
```
nit: there's an `exe` helper function somewhere you can use as a shortcut for this line

onur-ozkan committed:
download beta compiler toolchain in bootstrap if it doesn't yet exist

This is needed for when the shell scripts bypass python altogether and run the downloaded
bootstrap directly. Changes are mainly provided from @jyn514, I just fixed the review notes.

Signed-off-by: ozkanonur <work@onurozkan.dev>

ozkanonur commented:
@bors r=Mark-Simulacrum rollup=iffy

bors commented:
:pushpin: Commit eeec73244d2ab68e23b17b13f1163b254e597718 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum eeec73244d2ab68e23b17b13f1163b254e597718 -->
<!-- homu: {"type":"Approved","sha":"eeec73244d2ab68e23b17b13f1163b254e597718","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit eeec73244d2ab68e23b17b13f1163b254e597718 with merge ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba...
<!-- homu: {"type":"BuildStarted","head_sha":"eeec73244d2ab68e23b17b13f1163b254e597718","merge_sha":"ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4641939703/jobs/8216559970)
Approved by: Mark-Simulacrum
Pushing ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4641939703/jobs/8216559970"},"merge_sha":"ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba"} -->

rust-timer commented:
Finished benchmarking commit (ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba): [comparison URL](https://perf.rust-lang.org/compare.html?start=23ee2af2f7669d521ad39ef8a506a4fa3d37ea5a&end=ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=23ee2af2f7669d521ad39ef8a506a4fa3d37ea5a&end=ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=23ee2af2f7669d521ad39ef8a506a4fa3d37ea5a&end=ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [3.0%, 3.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=23ee2af2f7669d521ad39ef8a506a4fa3d37ea5a&end=ba86600c9452508a0ab05cd4bdc9af4c96f7d4ba&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
